cpp/src/round/round.cu

```diff
@@ -271,7 +271,10 @@ std::unique_ptr<column> round_with(column_view const& input,
                                out_view.template end<Type>(),
                                static_cast<Type>(0));
   } else {
-    Type const n = std::pow(10, scale_movement);
+    Type n = 10;
+    for (int i = 1; i < scale_movement; ++i) {
```

Comment:
This should use exponentiation-by-squaring for efficiency, and we should have a common implementation of that in libcudf. (We already have two implementations of exponentiation-by-squaring.) I have started work on this and will push to this PR when it's ready, but that might not happen today.

Reply:
Awesome Bradley. 
Would you point to the other two implementations? I was trying to look for them myself earlier today.

Reply:
One is in fixed point code and has a base that is known as a template parameter:

https://github.com/rapidsai/cudf/blob/7825790eac838e7a852d9b2429c64e78710cee28/cpp/include/cudf/fixed_point/fixed_point.hpp#L93

The other is in binary ops code and its base and exponent are both runtime parameters:

https://github.com/rapidsai/cudf/blob/7825790eac838e7a852d9b2429c64e78710cee28/cpp/src/binaryop/compiled/operation.cuh#L251

I am not sure how to best refactor this, but I have drafted some work locally (not yet pushed) that would add a file `cudf/detail/utilities/intpow.hpp` that centralizes this logic and exposes both a "constexpr base" and "runtime base" form of the function. I'll push this soon so it can be evaluated -- but there's some hangups I am seeing locally with include order and `cuda_runtime.h` macro conflicts (`__forceinline__`) with CCCL (resolved in libcudacxx 2.2.0).

Reply:
Note that we do _not_ have an intpow AST operator, because one has not been requested (to the best of my knowledge). It would go somewhere in here, but would need to have a different name like `INTPOW` to disambiguate it from the operators that are expected to return floating point values: https://github.com/rapidsai/cudf/blob/7825790eac838e7a852d9b2429c64e78710cee28/cpp/include/cudf/ast/detail/operators.hpp#L403

Reply:
The topic of integer powers was heavily discussed and analyzed in https://github.com/rapidsai/cudf/issues/10178.

Reply:
There are two more places where I think this bug might reoccur:
- unary math ops: https://github.com/rapidsai/cudf/blob/7825790eac838e7a852d9b2429c64e78710cee28/cpp/src/unary/math_ops.cu#L298
- unary cast ops: https://github.com/rapidsai/cudf/blob/7825790eac838e7a852d9b2429c64e78710cee28/cpp/src/unary/cast_ops.cu#L202

I'd love help writing some tests that fail for these cases.

Reply:
I'm starting work on a follow-up PR to fix these additional rescaling issues in #14242. I have a checklist there. This PR should be limited in scope to fixing _only_ the rounding issues, to minimize friction for this fix. I'd like to target refactoring requests to #14242 (aiming for 23.10) or a subsequent release (probably 23.12)

Reply:
See also https://github.com/rapidsai/cudf/issues/9346

Reply:
I opened https://github.com/rapidsai/cudf/issues/14243 to track future work on this.